docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,159 @@ the worker pool has a finite size and even with the full batching algorithm in t
+
+#### Determining Transaction Schema Versions
+
+We still need to retain the ability to read from the old transactions table, when checking if/when values written
+before the transactions2 migration occurred. Thus, we replace the old ``TransactionService`` (that only reads/writes
+from transactions1) with one that delegates between multiple TransactionServices (concretely, just V1 and V2 at this
+time) depending on the specific timestamps involved in calls.
+
+The mechanism for determining which ``TransactionService`` to use goes through the aforementioned
+``CoordinationService``. We agree on a ``TimestampPartitioningMap``, which is a mapping of timestamps to transaction
+schema versions. This is a range map, as we want to support rollbacks easily. The map also is guaranteed to span the
+ranges ``[1, +∞)`` and be connected; note that this does not mean future behaviour is fixed, since in practice this map
+is read together with a validity bound. Thus, even though the map will contain as a key some range ``[C, +∞)`` for a
+constant ``C``, it should be valid up to some point ``V > C`` - and behaviour at timestamps after ``V`` may subsequently
+be changed.
+
+Note that the ranges are based on the start timestamp. To give a concrete example, if our range map is
+``{[1, 5000) = 1, [5000, +∞) = 2}`` and we want to commit a transaction that started at timestamp 4990 and finished
+at 5010, it is still stored under schema version 1 (even if other transactions that started later may have already
+written to schema version 2 - if they started at 5000 and finished at 5001, for instance).
+
+We attempt to read the latest version of the range map, and if our timestamp falls within the validity bound, we
+retrieve the version. If it does not, we submit an identity transformation, which extends the validity bound of the
+current value. We then read the value again, and keep trying until we know what version to use. If we read a version
+we don't support yet, we throw (presumably, this is part of a rolling upgrade and should resolve itself soon).
+
+#### Changing Transaction Schema Versions
+
+AtlasDB clients continuously run a background thread that polls the runtime configuration. If that sets the
+target schema version to a specific version (which at time of writing can be 1 or 2), a transform will be run
+setting the transactions schema version going forward to that version.
+
+This is optional; if not configured, we will not attempt to run any transforms so the transaction schema version will
+remain what it is.
+
+### Live Migrations
+
+Transactions2 was written with some of the heaviest users of AtlasDB in mind. These are core Palantir services where
+shutdown upgrades are costly or even verboten, and we thus implemented a mechanism for performing upgrades without
+downtime. The key problem here involves reasoning about service nodes running different binary versions that are
+concurrently operating on an AtlasDB deployment.
```

Comment:
I believe the important distinction is 
> different binary versions that may be running different atlasdb versions.

This paragraph feels somewhat incomplete, maybe the bottom sections are sufficient

Reply:
Yeah, I will add that. The purpose of this paragraph is to argue the case for why live migrations for transactions are important, so I think it (or something to that effect) should stay - the bottom sections explain how you would do it but not why we want it.